Sources/Commands/SwiftRunTool.swift

```diff
@@ -0,0 +1,149 @@
+/*
+This source file is part of the Swift.org open source project
+
+Copyright 2015 - 2016 Apple Inc. and the Swift project authors
+Licensed under Apache License v2.0 with Runtime Library Exception
+
+See http://swift.org/LICENSE.txt for license information
+See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+import class Foundation.ProcessInfo
+
+import Basic
+import Build
+import Utility
+import PackageGraph
+import PackageModel
+
+import func POSIX.chdir
+import func POSIX.getcwd
+
+/// An enumeration of the errors that can be generated by the run tool.
+private enum RunError: Swift.Error {
+    /// The package manifest has no executable product.
+    case noExecutableFound
+
+    /// Could not find a specific executable in the package manifest.
+    case executableNotFound(String)
+
+    /// There are multiple executables and one must be chosen.
+    case multipleExecutables([String])
+}
+
+extension RunError: CustomStringConvertible {
+    var description: String {
+        switch self {
+        case .noExecutableFound:
+            return "no executable product in the package"
+        case .executableNotFound(let executable):
+            return "could not find executable product '\(executable)' in the package"
+        case .multipleExecutables(let executables):
+            let joinedExecutables = executables.joined(separator: ", ")
+            return "multiple executable products in the package. Use `swift run ` followed by one of: \(joinedExecutables)"
+        }
+    }
+}
+
+public class RunToolOptions: ToolOptions {
+    /// Returns the mode in with the tool command should run.
+    var mode: RunMode {
+        // If we got version option, just print the version and exit.
+        if shouldPrintVersion {
+            return .version
+        }
+
+        return .run
+    }
+    
+    /// If the executable product should be built before running.
+    var shouldBuild = true
+    
+    /// The executable product to run.
+    var executable: String?
+    
+    /// The arguments to pass to the executable.
+    var arguments: [String] = []
+}
+
+public enum RunMode {
+    case version
+    case run
+}
+
+/// swift-run tool namespace
+public class SwiftRunTool: SwiftTool<RunToolOptions> {
+
+   public convenience init(args: [String]) {
+       self.init(
+            toolName: "run",
+            usage: "[options] [executable [arguments ...]]",
+            overview: "Build and run an executable product",
+            args: args
+        )
+    }
+
+    override func runImpl() throws {
+        switch options.mode {
+        case .version:
+            print(Versioning.currentVersion.completeDisplayString)
+
+        case .run:
+            let plan = try buildPlan()
+
+            if options.shouldBuild {
+                try build(plan: plan, includingTests: false)
+            }
+
+            try run(findExecutable(in: plan))
+        }
+    }
+
+    /// Find executable path based on options.
```

Comment:
nit: The API guidelines recommend wording the doc comments in terms of what the method returns. Something like `Returns executable path ...` will be more appropriate.